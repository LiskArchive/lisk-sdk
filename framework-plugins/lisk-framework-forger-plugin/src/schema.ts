/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const forgerInfoSchema = {
	$id: '/forger/info',
	type: 'object',
	properties: {
		totalProducedBlocks: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		totalReceivedFees: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		totalReceivedRewards: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		votesReceived: {
			type: 'array',
			items: {
				type: 'object',
				properties: {
					address: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
			fieldNumber: 4,
			required: ['address', 'amount'],
		},
	},
	required: ['totalProducedBlocks', 'totalReceivedFees', 'totalReceivedRewards', 'votesReceived'],
};
