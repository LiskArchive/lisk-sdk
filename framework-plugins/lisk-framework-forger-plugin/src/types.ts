/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export interface Forger {
	readonly forging: boolean;
	readonly address: string;
}

export interface ForgerInfo {
	totalProducedBlocks: number;
	totalReceivedFees: bigint;
	totalReceivedRewards: bigint;
	votesReceived: Voters[];
}

export interface Voters {
	address: Buffer;
	amount: bigint;
}

export interface ForgetSyncInfo {
	syncUptoHeight: number;
}

export interface DPoSAccountJSON {
	pos: {
		validator: {
			username: string;
			totalStakeReceived: string;
			consecutiveMissedBlocks: number;
		};
	};
}
