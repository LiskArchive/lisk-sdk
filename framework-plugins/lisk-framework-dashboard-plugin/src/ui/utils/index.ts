/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { apiClient } from '@liskhq/lisk-client';
import { Block, Transaction } from '../types';

interface Config {
	applicationUrl: string;
	applicationName: string;
}

const configDevEnvValues: Config = {
	applicationUrl: 'ws://localhost:8080/ws',
	applicationName: 'Lisk',
};

const MAX_BLOCKS = 103 * 3;
const MAX_TRANSACTIONS = 150;

const sortByBlockHeight = (a: Block, b: Block) => b.header.height - a.header.height;

export const updateStatesOnNewBlock = (
	client: apiClient.APIClient,
	newBlockStr: string,
	blocks: Block[],
	confirmedTransactions: Transaction[],
	unconfirmedTransactions: Transaction[],
): {
	blocks: Block[];
	confirmedTransactions: Transaction[];
	unconfirmedTransactions: Transaction[];
} => {
	const newBlock = client.block.toJSON(client.block.decode(newBlockStr));
	let newBlocks = blocks;
	if (!blocks.find(b => b.header.height === newBlock.header.height)) {
		newBlocks = [newBlock, ...blocks].slice(0, MAX_BLOCKS) as Block[];
	}
	newBlocks.sort(sortByBlockHeight);

	for (const trs of newBlock.transactions) {
		confirmedTransactions.unshift(trs as unknown as Transaction);
	}
	const confirmedTransactionsIds = confirmedTransactions.map(t => t.id);
	const newUnconfirmedTransactions = unconfirmedTransactions
		.filter(t => !confirmedTransactionsIds.includes(t.id))
		.slice(0, MAX_TRANSACTIONS);

	const newConfirmedTransactions = confirmedTransactions.slice(0, MAX_TRANSACTIONS);

	return {
		blocks: newBlocks,
		confirmedTransactions: newConfirmedTransactions,
		unconfirmedTransactions: newUnconfirmedTransactions,
	};
};

export const updateStatesOnNewTransaction = (
	client: apiClient.APIClient,
	newTransactionStr: string,
	unconfirmedTransactions: Transaction[],
): Transaction[] => {
	const transaction = client.transaction.toJSON(client.transaction.decode(newTransactionStr));
	return [transaction, ...unconfirmedTransactions].slice(-1 * MAX_TRANSACTIONS) as Transaction[];
};

export const getConfig = async () => {
	if (process.env.NODE_ENV === 'development') {
		return configDevEnvValues;
	}

	const apiResponse = await fetch('/api/config');
	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
	const result: Config = await apiResponse.json();

	return result;
};
