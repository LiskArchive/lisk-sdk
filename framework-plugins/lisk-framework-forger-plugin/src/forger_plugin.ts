/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { Server } from 'http';
import * as fs from 'fs-extra';
import * as path from 'path';
import * as os from 'os';
import { getAddressFromPublicKey } from '@liskhq/lisk-cryptography';
import { codec } from '@liskhq/lisk-codec';
import { KVStore } from '@liskhq/lisk-db';
import {
	ActionsDefinition,
	BasePlugin,
	BaseChannel,
	EventsArray,
	EventInfoObject,
	PluginInfo,
} from 'lisk-framework';
import { constants as transactionConstants } from '@liskhq/lisk-transactions';
import { objects } from '@liskhq/lisk-utils';
import * as express from 'express';
import type { Express } from 'express';
import * as cors from 'cors';
import * as rateLimit from 'express-rate-limit';
import * as controllers from './controllers';
import * as middlewares from './middlewares';
import * as config from './defaults';
import { Forger, ForgerInfo, Options } from './types';
import { DB_KEY_FORGER_INFO } from './constants';
import { forgerInfoSchema } from './schema';

// eslint-disable-next-line
const packageJSON = require('../package.json');

export class ForgerPlugin extends BasePlugin {
	private _forgerPluginDB!: KVStore;
	private _server!: Server;
	private _app!: Express;
	private _channel!: BaseChannel;
	private _forgersList!: ReadonlyArray<Forger>;

	// eslint-disable-next-line @typescript-eslint/class-literal-property-style
	public static get alias(): string {
		return 'forger';
	}

	// eslint-disable-next-line @typescript-eslint/class-literal-property-style
	public static get info(): PluginInfo {
		return {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			author: packageJSON.author,
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			version: packageJSON.version,
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			name: packageJSON.name,
		};
	}

	// eslint-disable-next-line class-methods-use-this
	public get defaults(): object {
		return config.defaultConfig;
	}

	// eslint-disable-next-line class-methods-use-this
	public get events(): EventsArray {
		return [];
	}

	// eslint-disable-next-line class-methods-use-this
	public get actions(): ActionsDefinition {
		return {};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async load(channel: BaseChannel): Promise<void> {
		this._app = express();
		const options = objects.mergeDeep({}, config.defaultConfig.default, this.options) as Options;
		this._channel = channel;

		this._forgerPluginDB = await this._getDBInstance(options);

		// eslint-disable-next-line @typescript-eslint/no-misused-promises
		this._channel.once('app:ready', async () => {
			this._registerMiddlewares(options);
			this._registerControllers();
			this._registerAfterMiddlewares(options);
			this._subscribeToChannel();
			await this._setForgersInfo();
			this._server = this._app.listen(options.port, '0.0.0.0');
		});
	}

	public async unload(): Promise<void> {
		await new Promise((resolve, reject) => {
			this._server.close(err => {
				if (err) {
					reject(err);
					return;
				}
				resolve();
			});
		});

		await this._forgerPluginDB.close();
	}

	private _registerMiddlewares(options: Options): void {
		// Register middlewares
		this._app.use(cors(options.cors));
		this._app.use(express.json());
		this._app.use(rateLimit(options.limits));
		this._app.use(middlewares.whiteListMiddleware(options));
	}

	private _registerAfterMiddlewares(_options: Options): void {
		this._app.use(middlewares.errorMiddleware());
	}

	private _registerControllers(): void {
		this._app.get('/v1/hello', controllers.helloController(this._channel));
	}

	// eslint-disable-next-line class-methods-use-this
	private async _getDBInstance(options: Options, dbName = 'forger_plugin.db'): Promise<KVStore> {
		const resolvedPath = options.dataPath.replace('~', os.homedir());
		const dirPath = path.join(resolvedPath, dbName);
		await fs.ensureDir(dirPath);

		return new KVStore(dirPath);
	}

	private _subscribeToChannel(): void {
		// eslint-disable-next-line @typescript-eslint/no-misused-promises
		this._channel.subscribe('app:block:new', async (newBlock: EventInfoObject) => {
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			const { header: { generatorPublicKey }, payload } = this.codec.decodeBlock((newBlock.data as any).block);
			const forgerInfo = await this._getForgerInfo(generatorPublicKey);
			let totalProducedBlocks = forgerInfo.totalProducedBlocks;
			let totalReceivedRewards = forgerInfo.totalReceivedRewards;
			let totalReceivedFees = forgerInfo.totalReceivedFees;

			if (this._checkIfForgingDelegate(generatorPublicKey)) {
				totalProducedBlocks += 1;
				totalReceivedRewards = forgerInfo.totalReceivedRewards;
				totalReceivedFees = forgerInfo.totalReceivedFees;
			}

			const votesReceived = [...forgerInfo.votesReceived, payload.map(trx => trx.type === )]

			await this._setForgerInfo(generatorPublicKey, {
				totalProducedBlocks,
				totalReceivedFees,
				totalReceivedRewards,
				votesReceived,
			});
		});

		// this._channel.subscribe('app:block:delete', newBlock => {
		// 	console.log({ newBlock }, 'newBlock...................');
		// });
	}

	private _checkIfForgingDelegate(generatorPublicKey: string): boolean {
		const forgerAddress = getAddressFromPublicKey(Buffer.from(generatorPublicKey, 'base64')).toString('base64');

		return !this._forgersList.find(forger => forger.address === forgerAddress);
	}

	private async _setForgersInfo(): Promise<void> {
		try {
			const forgingDelegates = await this._channel.invoke<Forger[]>('app:getForgingStatusOfAllDelegates');
			this._forgersList = forgingDelegates.filter(forgers => forgers.forging);
		} catch (error) {
			throw new Error(`Action app:getForgingStatusOfAllDelegates failed with error: ${(error as Error).message}`);
		}
	}

	private async _getForgerInfo(generatorPublicKey: string): Promise<ForgerInfo> {
		const forgerInfo = await this._forgerPluginDB.get(`${DB_KEY_FORGER_INFO}:${generatorPublicKey}`);

		return codec.decode<ForgerInfo>(forgerInfoSchema, forgerInfo);
	}

	private async _setForgerInfo(generatorPublicKey: string, forgerInfo: ForgerInfo): Promise<void> {
		const encodedForgerInfo = codec.encode(forgerInfoSchema, forgerInfo);
		await this._forgerPluginDB.put(`${DB_KEY_FORGER_INFO}:${generatorPublicKey}`, encodedForgerInfo);
	}
}
