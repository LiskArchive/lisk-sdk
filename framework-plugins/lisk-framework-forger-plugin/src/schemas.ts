/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const forgerInfoSchema = {
	$id: '/forger/info',
	type: 'object',
	properties: {
		totalProducedBlocks: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		totalReceivedFees: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		totalReceivedRewards: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		stakeReceived: {
			type: 'array',
			fieldNumber: 4,
			items: {
				type: 'object',
				properties: {
					address: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
			required: ['address', 'amount'],
		},
	},
	required: ['totalProducedBlocks', 'totalReceivedFees', 'totalReceivedRewards', 'stakeReceived'],
};

export const forgerSyncSchema = {
	$id: '/forger/sync',
	type: 'object',
	required: ['syncUptoHeight'],
	properties: {
		syncUptoHeight: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
	},
};

export const posVoteParamsSchema = {
	$id: '/forger/pos/stake',
	type: 'object',
	required: ['stakes'],
	properties: {
		stakes: {
			type: 'array',
			minItems: 1,
			maxItems: 20,
			items: {
				type: 'object',
				required: ['validatorAddress', 'amount'],
				properties: {
					validatorAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
						minLength: 20,
						maxLength: 20,
					},
					amount: {
						dataType: 'sint64',
						fieldNumber: 2,
					},
				},
			},
			fieldNumber: 1,
		},
	},
};
