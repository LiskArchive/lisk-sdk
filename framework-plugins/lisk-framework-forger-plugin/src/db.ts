/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import * as createDebug from 'debug';
import { codec, db as liskDB } from 'lisk-sdk';
import * as os from 'os';
import { join } from 'path';
import { ensureDir } from 'fs-extra';
import { DB_KEY_FORGER_INFO, DB_KEY_FORGER_SYNC_INFO } from './constants';
import { forgerInfoSchema, forgerSyncSchema } from './schemas';
import { ForgerInfo, ForgetSyncInfo } from './types';

const debug = createDebug('plugin:forger:db');

const { Database } = liskDB;
type KVStore = liskDB.Database;

export const getDBInstance = async (
	dataPath: string,
	dbName = 'lisk-framework-forger-plugin.db',
): Promise<KVStore> => {
	const dirPath = join(dataPath.replace('~', os.homedir()), 'plugins/data', dbName);
	await ensureDir(dirPath);

	return new Database(dirPath);
};

export const getForgerSyncInfo = async (db: KVStore): Promise<ForgetSyncInfo> => {
	try {
		const encodedSyncInfo = await db.get(DB_KEY_FORGER_SYNC_INFO);
		return codec.decode<ForgetSyncInfo>(forgerSyncSchema, encodedSyncInfo);
	} catch (error) {
		debug('Forger sync info does not exists');
		return {
			syncUptoHeight: 0,
		};
	}
};

export const setForgerSyncInfo = async (db: KVStore, blockHeight: number): Promise<void> => {
	const encodedSyncInfo = codec.encode(forgerSyncSchema, { syncUptoHeight: blockHeight });
	await db.set(DB_KEY_FORGER_SYNC_INFO, encodedSyncInfo);
};

export const setForgerInfo = async (
	db: KVStore,
	forgerAddress: string,
	forgerInfo: ForgerInfo,
): Promise<void> => {
	const encodedForgerInfo = codec.encode(forgerInfoSchema, forgerInfo);
	await db.set(
		Buffer.concat([DB_KEY_FORGER_INFO, Buffer.from(`:${forgerAddress}`, 'utf8')]),
		encodedForgerInfo,
	);
};

export const getForgerInfo = async (db: KVStore, forgerAddress: string): Promise<ForgerInfo> => {
	let forgerInfo;
	try {
		forgerInfo = await db.get(
			Buffer.concat([DB_KEY_FORGER_INFO, Buffer.from(`:${forgerAddress}`, 'utf8')]),
		);
	} catch (error) {
		debug(`Forger info does not exists for validator: ${forgerAddress}`);
		return {
			totalProducedBlocks: 0,
			totalReceivedFees: BigInt(0),
			totalReceivedRewards: BigInt(0),
			votesReceived: [],
		};
	}

	return codec.decode<ForgerInfo>(forgerInfoSchema, forgerInfo);
};
