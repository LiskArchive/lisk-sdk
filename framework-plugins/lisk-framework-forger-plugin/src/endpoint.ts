/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import {
	BasePlugin,
	BasePluginEndpoint,
	PluginEndpointContext,
	db as liskDB,
	cryptography,
} from 'lisk-sdk';
import { getForgerInfo } from './db';
import { Forger } from './types';

interface Staker {
	readonly address: string;
	readonly username: string;
	readonly totalStakeReceived: string;
	readonly stakers: {
		readonly address: string;
		readonly amount: string;
	}[];
}

interface ForgerInfo extends Forger {
	readonly username: string;
	readonly totalReceivedFees: string;
	readonly totalReceivedRewards: string;
	readonly totalProducedBlocks: number;
	readonly totalStakeReceived: string;
	readonly consecutiveMissedBlocks: number;
}

interface Validator {
	name: string;
	totalStakeReceived: string;
	selfStake: string;
	lastGeneratedHeight: number;
	isBanned: boolean;
	pomHeights: ReadonlyArray<number>;
	consecutiveMissedBlocks: number;
}

export class Endpoint extends BasePluginEndpoint {
	private _client!: BasePlugin['apiClient'];
	private _db!: liskDB.Database;

	public init(db: liskDB.Database, apiClient: BasePlugin['apiClient']) {
		this._db = db;
		this._client = apiClient;
	}

	public async getStakers(_context: PluginEndpointContext): Promise<Staker[]> {
		const forgersList = await this._client.invoke<Forger[]>('app_getForgingStatus');
		const forgerAccounts = [];
		for (const validator of forgersList) {
			const res = await this._client.invoke<Validator>('pos_getValidator', {
				address: validator.address,
			});
			forgerAccounts.push({
				...res,
				address: validator.address,
			});
		}

		const result: Staker[] = [];
		for (const account of forgerAccounts) {
			const forgerInfo = await getForgerInfo(
				this._db,
				Buffer.from(account.address, 'hex').toString('binary'),
			);

			result.push({
				address: account.address,
				username: account.name,
				totalStakeReceived: account.totalStakeReceived,
				stakers: forgerInfo.stakeReceived.map(stake => ({
					address: cryptography.address.getLisk32AddressFromAddress(stake.address),
					amount: stake.amount.toString(),
				})),
			});
		}

		return result;
	}

	public async getForgingInfo(_context: PluginEndpointContext): Promise<ForgerInfo[]> {
		const forgersList = await this._client.invoke<ReadonlyArray<Forger>>('app_getForgingStatus');
		const forgerAccounts = [];
		for (const validator of forgersList) {
			const res = await this._client.invoke<Validator>('pos_getValidator', {
				address: validator.address,
			});
			forgerAccounts.push({
				...res,
				address: validator.address,
			});
		}
		const data: ForgerInfo[] = [];
		for (const forgerAccount of forgerAccounts) {
			const forgerAddressBinary = Buffer.from(forgerAccount.address, 'hex').toString('binary');
			const forgerInfo = await getForgerInfo(this._db, forgerAddressBinary);
			const forger = forgersList.find(aForger => aForger.address === forgerAccount.address);

			if (forger) {
				data.push({
					...forger,
					username: forgerAccount.name,
					totalReceivedFees: forgerInfo.totalReceivedFees.toString(),
					totalReceivedRewards: forgerInfo.totalReceivedRewards.toString(),
					totalProducedBlocks: forgerInfo.totalProducedBlocks,
					totalStakeReceived: forgerAccount.totalStakeReceived,
					consecutiveMissedBlocks: forgerAccount.consecutiveMissedBlocks,
				});
			}
		}

		return data;
	}
}
