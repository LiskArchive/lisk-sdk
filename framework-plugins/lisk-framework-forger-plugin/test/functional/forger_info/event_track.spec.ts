/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { testing, PartialApplicationConfig } from 'lisk-sdk';
import { getForgerInfoByAddress, getForgerPlugin, waitTill } from '../../utils/application';
import { getRandomAccount } from '../../utils/accounts';
import { createTransferTransaction, createStakeTransaction } from '../../utils/transactions';
import { ForgerPlugin } from '../../../src';

describe('Forger Info', () => {
	let appEnv: testing.ApplicationEnv;
	let accountNonce = 0;
	let chainID: Buffer;

	beforeAll(async () => {
		const rootPath = '~/.lisk/forger-plugin';
		const config = {
			rootPath,
			label: 'event_track_functional',
		} as PartialApplicationConfig;

		appEnv = testing.createDefaultApplicationEnv({
			config,
			plugins: [new ForgerPlugin()],
		});
		await appEnv.startApplication();
		// The test application generates a dynamic genesis block so we need to get the networkID like this
		chainID = appEnv.chainID;
	});

	afterAll(async () => {
		// eslint-disable-next-line @typescript-eslint/no-empty-function
		jest.spyOn(process, 'exit').mockImplementation((() => {}) as never);
		await appEnv.stopApplication();
	});

	describe('New Block', () => {
		it('should save forger info after new block', async () => {
			// Arrange
			const forgerPluginInstance = getForgerPlugin(appEnv.application);

			// Act
			const { generatorAddress } = appEnv.lastBlock.header;
			const forgerInfo = await getForgerInfoByAddress(
				forgerPluginInstance,
				generatorAddress.toString('binary'),
			);

			// Assert
			expect(forgerInfo).toMatchSnapshot();
		});

		it('should save forger info with received fees if transactions included in new block', async () => {
			// Arrange
			const forgerPluginInstance = getForgerPlugin(appEnv.application);
			const account = getRandomAccount();
			const transaction = createTransferTransaction({
				amount: '2',
				recipientAddress: account.address,
				fee: '0.3',
				nonce: accountNonce,
				chainID,
			});
			accountNonce += 1;

			await appEnv.ipcClient.invoke('txpool_postTransaction', {
				transaction: transaction.getBytes().toString('hex'),
			});
			await appEnv.waitNBlocks(1);

			const {
				header: { generatorAddress },
			} = appEnv.lastBlock;
			const forgerInfo = await getForgerInfoByAddress(
				forgerPluginInstance,
				generatorAddress.toString('binary'),
			);

			// Assert
			expect(forgerInfo).toMatchSnapshot();
		});

		describe('Stake transactions', () => {
			it('should save forger info with stakes received in new block', async () => {
				// Arrange
				const forgerPluginInstance = getForgerPlugin(appEnv.application);
				const [forgingValidatorAddress] = forgerPluginInstance['_forgersList'].entries()[0];
				const transaction1 = createStakeTransaction({
					amount: '10',
					recipientAddress: forgingValidatorAddress.toString('hex'),
					fee: '0.3',
					nonce: accountNonce,
					chainID,
				});
				accountNonce += 1;

				await appEnv.ipcClient.invoke('txpool_postTransaction', {
					transaction: transaction1.getBytes().toString('hex'),
				});
				await appEnv.waitNBlocks(1);
				await waitTill(200);

				const forgerInfo = await getForgerInfoByAddress(
					forgerPluginInstance,
					forgingValidatorAddress.toString('binary'),
				);
				// Assert
				expect(forgerInfo).toMatchSnapshot();
				expect(forgerInfo.stakeReceived[0].amount).toEqual(BigInt(1000000000));
			});

			it('should update forger info with multiple stakes received for same validator in new block', async () => {
				// Arrange
				const forgerPluginInstance = getForgerPlugin(appEnv.application);
				const [forgingValidatorAddress] = forgerPluginInstance['_forgersList'].entries()[0];
				const transaction1 = createStakeTransaction({
					amount: '10',
					recipientAddress: forgingValidatorAddress.toString('hex'),
					fee: '0.3',
					nonce: accountNonce,
					chainID,
				});
				accountNonce += 1;
				const transaction2 = createStakeTransaction({
					amount: '50',
					recipientAddress: forgingValidatorAddress.toString('hex'),
					fee: '0.3',
					nonce: accountNonce,
					chainID,
				});
				accountNonce += 1;

				await appEnv.ipcClient.invoke('txpool_postTransaction', {
					transaction: transaction1.getBytes().toString('hex'),
				});
				await appEnv.ipcClient.invoke('txpool_postTransaction', {
					transaction: transaction2.getBytes().toString('hex'),
				});
				await appEnv.waitNBlocks(1);
				await waitTill(200);

				const forgerInfo = await getForgerInfoByAddress(
					forgerPluginInstance,
					forgingValidatorAddress.toString('binary'),
				);
				// Assert
				expect(forgerInfo).toMatchSnapshot();
				expect(forgerInfo.stakeReceived[0].amount).toEqual(BigInt(7000000000));
			});

			it('should update forger info with upvote and downvote for same validator in new block', async () => {
				// Arrange
				const forgerPluginInstance = getForgerPlugin(appEnv.application);
				const [forgingValidatorAddress] = forgerPluginInstance['_forgersList'].entries()[0];
				const transaction1 = createStakeTransaction({
					amount: '-50',
					recipientAddress: forgingValidatorAddress.toString('hex'),
					fee: '0.3',
					nonce: accountNonce,
					chainID,
				});
				accountNonce += 1;
				const transaction2 = createStakeTransaction({
					amount: '+10',
					recipientAddress: forgingValidatorAddress.toString('hex'),
					fee: '0.3',
					nonce: accountNonce,
					chainID,
				});
				accountNonce += 1;

				await appEnv.ipcClient.invoke('txpool_postTransaction', {
					transaction: transaction1.getBytes().toString('hex'),
				});
				await appEnv.ipcClient.invoke('txpool_postTransaction', {
					transaction: transaction2.getBytes().toString('hex'),
				});
				await appEnv.waitNBlocks(1);
				await waitTill(200);

				const forgerInfo = await getForgerInfoByAddress(
					forgerPluginInstance,
					forgingValidatorAddress.toString('binary'),
				);
				// Assert
				expect(forgerInfo).toMatchSnapshot();
				expect(forgerInfo.stakeReceived[0].amount).toEqual(BigInt(3000000000));
			});

			it('should update forger info with stakers info and remove when amount becomes zero', async () => {
				// Arrange
				const forgerPluginInstance = getForgerPlugin(appEnv.application);
				const [forgingValidatorAddress1] = forgerPluginInstance['_forgersList'].entries()[0];
				const [forgingValidatorAddress2] = forgerPluginInstance['_forgersList'].entries()[1];
				const transaction1 = createStakeTransaction({
					amount: '-30',
					recipientAddress: forgingValidatorAddress1.toString('hex'),
					fee: '0.3',
					nonce: accountNonce,
					chainID,
				});
				accountNonce += 1;
				const transaction2 = createStakeTransaction({
					amount: '20',
					recipientAddress: forgingValidatorAddress2.toString('hex'),
					fee: '0.3',
					nonce: accountNonce,
					chainID,
				});
				accountNonce += 1;

				await appEnv.ipcClient.invoke('txpool_postTransaction', {
					transaction: transaction1.getBytes().toString('hex'),
				});
				await appEnv.ipcClient.invoke('txpool_postTransaction', {
					transaction: transaction2.getBytes().toString('hex'),
				});
				await appEnv.waitNBlocks(1);
				await waitTill(200);

				const forgerInfo1 = await getForgerInfoByAddress(
					forgerPluginInstance,
					forgingValidatorAddress1.toString('binary'),
				);
				const forgerInfo2 = await getForgerInfoByAddress(
					forgerPluginInstance,
					forgingValidatorAddress2.toString('binary'),
				);
				// Assert
				expect(forgerInfo1).toMatchSnapshot();
				expect(forgerInfo1.stakeReceived).toBeEmpty();
				expect(forgerInfo2).toMatchSnapshot();
				expect(forgerInfo2.stakeReceived[0].amount).toEqual(BigInt(2000000000));
			});
		});
	});

	describe('Delete Block', () => {
		it('should update forger info after delete block', async () => {
			// Arrange
			const { generatorAddress } = appEnv.lastBlock.header;
			const forgerPluginInstance = getForgerPlugin(appEnv.application);
			await appEnv.application['_node']['_processor'].deleteLastBlock();

			// Act
			await waitTill(50);
			const forgerInfo = await getForgerInfoByAddress(
				forgerPluginInstance,
				generatorAddress.toString('binary'),
			);

			// Asserts
			expect(forgerInfo).toMatchSnapshot();
		});
	});
});
