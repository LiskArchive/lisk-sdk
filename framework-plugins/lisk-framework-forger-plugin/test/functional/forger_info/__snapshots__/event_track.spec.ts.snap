// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forger Info Delete Block should update forger info after delete block 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [],
}
`;

exports[`Forger Info New Block Vote transactions should save forger info with stakes received in new block 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [
    Object {
      "address": Object {
        "data": Array [
          208,
          70,
          153,
          229,
          124,
          74,
          56,
          70,
          201,
          136,
          243,
          193,
          83,
          6,
          121,
          111,
          142,
          174,
          92,
          28,
        ],
        "type": "Buffer",
      },
      "amount": 1000000000n,
    },
  ],
}
`;

exports[`Forger Info New Block Vote transactions should update forger info with multiple stakes received for same validator in new block 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [
    Object {
      "address": Object {
        "data": Array [
          208,
          70,
          153,
          229,
          124,
          74,
          56,
          70,
          201,
          136,
          243,
          193,
          83,
          6,
          121,
          111,
          142,
          174,
          92,
          28,
        ],
        "type": "Buffer",
      },
      "amount": 7000000000n,
    },
  ],
}
`;

exports[`Forger Info New Block Vote transactions should update forger info with upvote and downvote for same validator in new block 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [
    Object {
      "address": Object {
        "data": Array [
          208,
          70,
          153,
          229,
          124,
          74,
          56,
          70,
          201,
          136,
          243,
          193,
          83,
          6,
          121,
          111,
          142,
          174,
          92,
          28,
        ],
        "type": "Buffer",
      },
      "amount": 3000000000n,
    },
  ],
}
`;

exports[`Forger Info New Block Vote transactions should update forger info with stakers info and remove when amount becomes zero 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [],
}
`;

exports[`Forger Info New Block Vote transactions should update forger info with stakers info and remove when amount becomes zero 2`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [
    Object {
      "address": Object {
        "data": Array [
          208,
          70,
          153,
          229,
          124,
          74,
          56,
          70,
          201,
          136,
          243,
          193,
          83,
          6,
          121,
          111,
          142,
          174,
          92,
          28,
        ],
        "type": "Buffer",
      },
      "amount": 2000000000n,
    },
  ],
}
`;

exports[`Forger Info New Block should save forger info after new block 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [],
}
`;

exports[`Forger Info New Block should save forger info with received fees if transactions included in new block 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [],
}
`;
